File: windowsApp/DockhandWindowsApp/Models/Protocol.cs
using System.Text.Json.Serialization;

namespace DockhandWindowsApp.Models;

// Mirrors windowsApp/backend/src/main/kotlin/com/mrnrod45/dockhand/windows/Protocol.kt

public record Request(
    [property: JsonPropertyName("id")] int Id,
    [property: JsonPropertyName("method")] string Method,
    [property: JsonPropertyName("params")] object Params
);

public record Response(
    [property: JsonPropertyName("id")] int Id,
    [property: JsonPropertyName("result")] System.Text.Json.JsonElement? Result,
    [property: JsonPropertyName("error")] string? Error
);

public record Event(
    [property: JsonPropertyName("event")] string EventName,
    [property: JsonPropertyName("message")] string Message
);

public record UsbDeviceInfo(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("vendorId")] int VendorId,
    [property: JsonPropertyName("productId")] int ProductId
);

File: windowsApp/backend/src/main/kotlin/com/mrnrod45/dockhand/windows/Protocol.kt
package com.mrnrod45.dockhand.windows

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * Minimal newline-delimited JSON-RPC protocol between this JVM backend and the
 * WinUI 3 C# frontend, which spawns this process as a subprocess and talks to
 * it over a loopback TCP socket. See README.md for the full method list.
 */
@Serializable
data class Request(
    val id: Int,
    val method: String,
    val params: JsonObject = JsonObject(emptyMap())
)

@Serializable
data class Response(
    val id: Int,
    val result: JsonElement? = null,
    val error: String? = null
)

/** One-way push message (log lines during an upload/inject/split operation). */
@Serializable
data class Event(
    val event: String,
    val message: String
)
